mailet/standard/src/main/java/org/apache/james/transport/matchers/dlp/DlpDomainRules.java

```diff
@@ -0,0 +1,203 @@
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.matchers.dlp;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import javax.mail.Address;
+import javax.mail.MessagingException;
+import javax.mail.internet.MimeMessage;
+
+import org.apache.james.core.MailAddress;
+import org.apache.james.dlp.api.DLPConfigurationItem;
+import org.apache.james.util.OptionalUtils;
+import org.apache.mailet.Mail;
+
+import com.github.fge.lambdas.predicates.ThrowingPredicate;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
+public class DlpDomainRules {
+
+    @VisibleForTesting static DlpDomainRules matchNothing() {
+        return DlpDomainRules.of(new Rule(DLPConfigurationItem.Id.of("always false"), (mail) -> false));
+    }
+
+    @VisibleForTesting static DlpDomainRules matchAll() {
+        return DlpDomainRules.of(new Rule(DLPConfigurationItem.Id.of("always true"), (mail) -> true));
+    }
+
+    private static DlpDomainRules of(Rule rule) {
+        return new DlpDomainRules(ImmutableList.of(rule));
+    }
+
+    public static DlpDomainRulesBuilder builder() {
+        return new DlpDomainRulesBuilder();
+    }
+
+    static class Rule {
+
+        interface MatcherFunction extends ThrowingPredicate<Mail> { }
+
+        private final DLPConfigurationItem.Id id;
+        private final MatcherFunction matcher;
+
+        public Rule(DLPConfigurationItem.Id id, MatcherFunction matcher) {
+            this.id = id;
+            this.matcher = matcher;
+        }
+
+        public DLPConfigurationItem.Id id() {
+            return id;
+        }
+
+        public boolean match(Mail mail) {
+            return matcher.test(mail);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (o instanceof Rule) {
+                Rule other = (Rule) o;
+                return Objects.equals(id, other.id) &&
+                    Objects.equals(matcher, other.matcher);
+            }
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(id, matcher);
+        }
+
+    }
+
+    public static class DlpDomainRulesBuilder {
```

Comment:
This builder is not about building a DlpDomainRules but more about generating the rules.

I think having a factory for rule would be more straightforward, and easier to understand.

Reply:
What difference do you make between building and generating ?
The good thing about this design is that Rule is almost impossible to create outside DlpDomainRules.

Reply:
What difference do you make between building and generating ?
The goal of this design is to hide the Rule object as much as possible as it doesn't make sense outside DlpDomainRules.